src/eh1/io.rs

```diff
@@ -0,0 +1,617 @@
+            response,
+            expected_err: None,
+        }
+    }
+
+    /// Create a flush transaction
+    pub fn flush() -> Transaction {
+        Transaction {
+            expected_mode: Mode::Flush,
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Create a seek transaction
+    pub fn seek(pos: SeekFrom, ret_offset: u64) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Seek(pos),
+            expected_data: Vec::new(),
+            response: ret_offset.to_be_bytes().to_vec(),
+            expected_err: None,
+        }
+    }
+
+    /// Create a write ready transaction
+    pub fn write_ready(ready: bool) -> Transaction {
+        Transaction {
+            expected_mode: Mode::WriteReady(ready),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Create a read ready transaction
+    pub fn read_ready(ready: bool) -> Transaction {
+        Transaction {
+            expected_mode: Mode::ReadReady(ready),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Create a fill buffer transaction
+    pub fn fill_buf(response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::FillBuff,
+            expected_data: Vec::new(),
+            response,
+            expected_err: None,
+        }
+    }
+
+    /// Create a consume transaction
+    pub fn consume(consumed: usize) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Consume(consumed),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Add an error return to a transaction
+    ///
+    /// This is used to mock failure behaviors.
+    ///
+    /// Note: When attaching this to a read transaction, the response in the
+    /// expectation will not actually be written to the buffer.
+    pub fn with_error(mut self, error: ErrorKind) -> Self {
+        self.expected_err = Some(error);
+        self
+    }
+}
+
+/// Mock IO implementation
+pub type Mock = Generic<Transaction>;
+
+impl ErrorType for Mock {
+    type Error = ErrorKind;
+}
+
+impl Write for Mock {
+    fn write(&mut self, buffer: &[u8]) -> Result<usize, Self::Error> {
+        let w = self.next().expect("no expectation for io::write call");
+        assert_eq!(w.expected_mode, Mode::Write, "io::write unexpected mode");
+        assert_eq!(
+            &w.expected_data, &buffer,
+            "io::write data does not match expectation"
+        );
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(buffer.len()),
+        }
+    }
+
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        let w = self.next().expect("no expectation for io::flush call");
```

Comment:
The letter `w` was probably chosen in the `write` method to stand for "write". But in all other impls it seems a bit weird.

Maybe use `t` or `transaction` instead? (Here, and in all other impls as well.)